### ISSUE TITLE
Add storage folding support to buffers

### ISSUE THREAD

**Issue by dsharlet** [halide/Halide#8035]
In another project, I found it pretty useful and interesting to support fold factors on buffers directly, as opposed to adding modulo operations around coordinates in the code itself. Instead of dimensions of buffers being {min, extent, stride}, they would become {min, extent, stride, fold_factor}. The address computation would go from something like:
```
addr_of_x = base + sum((x_d - min_d) * stride_d)
```
to:
```
addr_of_x = base + sum(((x_d - min_d) % fold_factor_d) * stride_d)
```

Of course, some care would need to be taken for performance. I think this would need something similar to strides: the default stride of dimension 0 is 1, so vectorization doesn't turn into a gather by default. Similarly, the fold factor of dimension 0 should probably default to some value indicating "no folding".

However, for most dimensions, folding is nearly free, due to being computed in outer loops.

@abadams and @steven-johnson both noted that `halide_dimension_t` already has a flags field that is currently unused that could be used to support this parameter: https://github.com/halide/Halide/blob/22f9bb9247b3e384bbd9d8e7ff96501a29b49265/src/runtime/HalideRuntime.h#L1470-L1471

It might be useful and worth adding this to Halide. It would make it easier to plug Halide pipelines into other systems that work with storage folded buffers. Currently, the only option to do this would be to add a new parameter to the pipeline, and fold the dimension manually, by modding the coordinates in the code explicitly. However, doing this makes the bounds of the pipeline harder to reason about.

**Comment by abadams** [comment]
Worth keeping in mind that on many platforms you can fake this with memory mapping tricks (mapping multiple addresses to the same page). The buffer allocator needs to do the fancy mapping, but the buffer consumer doesn't need to know about folding or do any modulos.

**Comment by dsharlet** [comment]
Yes, although that approach is quite limiting (you can only fold on page aligned boundaries).

**Comment by abadams** [comment]
I raised it because it's the only method I know of that doesn't require any modifications on the consumer side, so we could potentially use it for existing libraries where we have no ability to modify the source.
